### DISCUSSION TITLE
tokio compatibility

### DISCUSSION THREAD

**Discussion by bernardoaraujor** [antonilol/rust-bitcoincore-zmq#3] [Ideas]
I working on this [PR to `braidpool`](https://github.com/braidpool/braidpool/pull/54).

Since `braidpool` project is using `tokio`, I forked `rust-bitcoincore-zmq` into [`tokio-bitcoincore-zmq`](https://github.com/bernardoaraujor/tokio-bitcoincore-zmq).

I was wondering if this is a desirable feature for the upstream crate?

**Comment by antonilol** [comment]
great idea! i have tried in the past to make this library work with async because zmq is mostly I/O and technically does not need one thread per subscriber, but just smashing an async runtime (i tried tokio) on it did not work (obviously, but i was exploring async Rust for the first time then) as it blocked the executor. i just looked around again and found https://github.com/rdelfin/async-zmq, this also does not force a specific async runtime upon users.
i will try around with this a bit and look how it fits in with the synchronous code, because i dont want to remove that, thinking about an 'async' feature flag, as it comes with extra dependencies not needed otherwise

**Reply by bernardoaraujor** [reply]
feature gating it could be cool!

you can check [my commit](https://github.com/bernardoaraujor/tokio-bitcoincore-zmq/commit/60b479ecc9cea126af28d8a33ac7de567d31c0fa) as a reference for the changes I applied. Basically I replaced the `callback` mechanism with the `mpsc` sender on `subscribe_internal`. Although I'm sure there might be more elegant ways to go about this, as this is a really `tokio`-biased approach.

**Reply by antonilol** [reply]
did you test it with less executor threads than endpoints? i believe it will lock up because `subscribe_internal` is blocking (it uses a `loop`). 
if you for example want to open a tcp socket in async rust you have to use an async tcp library ([tokio has them](https://docs.rs/tokio/latest/tokio/net/struct.TcpStream.html)), not a sync one line [std::net::TcpStream](https://doc.rust-lang.org/std/net/struct.TcpStream.html) as it will block a thread of the executor.

(async_zmq fits really well in my existing code, maybe i can push some testing code already today or tomorrow, will push this to a new branch called 'async')

**Reply by antonilol** [reply]
just implemented it, feel free to review for anything like questions, feature requests or maybe i missed something
#4

**Reply by bernardoaraujor** [reply]
didn't review everything, but seems to be on the right direction

I'm already able to use `subscribe_async` and it's working as expected, so that fulfills the needs for `braidpool`.

therefore I'm archiving my `tokio`-compatible fork in favor of this upstream project.

**Reply by antonilol** [reply]
> I'm already able to use `subscribe_async` and it's working as expected, so that fulfills the needs for `braidpool`.

good to hear, just a heads up, `subscribe_async` has been renamed to `subscribe_single_async` for consistency with the existing subscribe functions

**Reply by antonilol** [reply]
published version 1.3.0 to crates.io, the feature gates functions and structs are not showing on docs.rs, will fix this soon
